### PR TITLE
Add cmake build to travis #2

### PR DESCRIPTION
**Pull request by wilsonianb**
based on https://github.com/vinniefalco/Beast/pull/29

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -20,7 +20,28 @@ else()
...
+      "${CMAKE_CXX_FLAGS} -fsanitize=undefined -fno-omit-frame-pointer")
+    set (CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -fsanitize=undefined")
+    set(CMAKE_BUILD_TYPE RELWITHDEBINFO)
+elseif (${VARIANT} STREQUAL "debug")
+    set(CMAKE_BUILD_TYPE DEBUG)
+elseif (${VARIANT} STREQUAL "release")
+    set(CMAKE_BUILD_TYPE RELEASE)
 endif()
 
 message ("cxx Flags: " ${CMAKE_CXX_FLAGS})
```
Comment (seelabs) [inline CMakeLists.txt]: We might want to remove that message. It looks like debug code that was accidentally left in.

File: CMakeLists.txt [hunk 2]
```diff
@@ -20,7 +20,28 @@ else()
...
+endif()
+
+if (${VARIANT} STREQUAL "coverage")
+    set(CMAKE_CXX_FLAGS
+      "${CMAKE_CXX_FLAGS} -fprofile-arcs -ftest-coverage")
+    set(CMAKE_BUILD_TYPE RELWITHDEBINFO)
+    set (CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -lgcov")
+elseif (${VARIANT} STREQUAL "asan")
+    set(CMAKE_CXX_FLAGS
+      "${CMAKE_CXX_FLAGS} -fsanitize=address -fno-omit-frame-pointer")
```
Comment (seelabs) [inline CMakeLists.txt]: Not for today, but we might want to clean this up to use `add_compile_options` instead.

**Comment by seelabs** [comment]
:+1:

File: CMakeLists.txt [hunk 3]
```diff
@@ -20,7 +20,28 @@ else()
     find_package(Threads)
 
     set(CMAKE_CXX_FLAGS
-      "${CMAKE_CXX_FLAGS} -g -std=c++11 -Wall -Wpedantic")
+      "${CMAKE_CXX_FLAGS} -std=c++11 -Wall -Wpedantic")
+endif()
+
+if (${VARIANT} STREQUAL "coverage")
```
Comment (gebart) [inline CMakeLists.txt]: This line (and the subsequent similar elseifs) fails if cmake is run without `-DVARIANT=anything`.

```
~/src/beast/build % cmake ..
-- The C compiler identification is GNU 5.4.0
-- The CXX compiler identification is GNU 5.4.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Boost version: 1.61.0
-- Found the following Boost libraries:
--   coroutine
--   context
--   thread
--   filesystem
--   program_options
--   system
--   chrono
--   date_time
--   atomic
-- Check if compiler accepts -pthread
-- Check if compiler accepts -pthread - yes
CMake Error at CMakeLists.txt:26 (if):
  if given arguments:

    "STREQUAL" "coverage"

  Unknown arguments specified


-- Configuring incomplete, errors occurred!
```

Comment (vinniefalco) [inline CMakeLists.txt]: Not sure how this popped up after the code was merged but its being looked into, thanks.

**Pull request closed without merging**